packages/react-compose/src/compose.ts

```diff
@@ -1,37 +1,66 @@
 import * as React from 'react';
+import {
+  ComponentWithAs,
+  ComposedComponent,
+  ComposeOptions,
+  Input,
+  InputComposeComponent,
+  ComposePreparedOptions,
+} from './types';
+import { mergeComposeOptions } from './mergeComposeOptions';
 
-import { ComponentWithAs, ComposedComponent, ComposeOptions, Input, InputComposeComponent } from './types';
-import { mergeComposeOptions, wasComposedPreviously } from './utils';
-
-function compose<ElementType extends React.ElementType, InputProps, InputStylesProps, ParentProps, ParentStylesProps>(
-  input: Input<ElementType, InputProps>,
-  inputOptions: ComposeOptions<InputProps, InputStylesProps, ParentProps, ParentStylesProps> = {},
+/**
+ * The `compose` utility allows React components to be created using a set of default config settings. Resulting
+ * component can be "recomposed" by calling `compose` with them as the first argument, attaching config overrides
+ * in the second argument. Settings are merged correctly and provided to the component render function.
+ *
+ * @param input - render function `(props, ref, config) => JSX.Element` or previously composed component.
+ * @param inputOptions - options for rendering the component.
+ */
+function compose<
+  TElementType extends React.ElementType,
+  TProps = {},
```

Comment:
Would be better to stage the generics rename as a separate PR, just for easier reviewing :)